### ISSUE TITLE
BCEWithLogitsLoss computes wrong gradient?

### ISSUE THREAD

**Issue by amrsharaf** [pytorch/pytorch#2298]
Here are two identical pieces of code using BCELoss and BCEWithLogitsLoss, the loss is similar, but the gradient is different, the correct gradient should be 0.5:

```
def main():
    print('BCELoss():')
    criterion = nn.BCELoss()
    w = Variable(torch.Tensor([0.0]), requires_grad=True)
    x = Variable(torch.Tensor([1]), requires_grad=True)
    s = w * x
    loss = criterion(F.sigmoid(s), Variable(torch.Tensor([0.0])))
    print('loss: ', loss)
    loss.backward()
    print('w.grad: ', w.grad)
    print('==============')
    print('BCEWithLogitsLoss():')
    criterion = nn.BCEWithLogitsLoss()
    w = Variable(torch.Tensor([0.0]), requires_grad=True)
    x = Variable(torch.Tensor([1]), requires_grad=True)
    s = w * x
    loss = criterion(s, Variable(torch.Tensor([0.0])))
    print('loss: ', loss)
    loss.backward()
    print('w.grad: ', w.grad)
    print('==============')
```

output:

> BCELoss():
> loss:  Variable containing:
>  0.6931
> [torch.FloatTensor of size 1]
> 
> w.grad:  Variable containing:
>  0.5000
> [torch.FloatTensor of size 1]
> 
> ==============
> BCEWithLogitsLoss():
> loss:  Variable containing:
>  0.6931
> [torch.FloatTensor of size 1]
> 
> w.grad:  Variable containing:
>  0
> [torch.FloatTensor of size 1]

**Comment by amrsharaf** [comment]
CC: @alykhantejani

**Comment by alykhantejani** [comment]
Hi @amrsharaf are you using master? This was fixed and merged here: https://github.com/pytorch/pytorch/pull/2195

**Comment by amrsharaf** [comment]
Thanks @alykhantejani I indeed haven't compiled in few days, I'll pull the fix and recompile!
